File: bindings/python/tests/test_render.py
"""The manual sink: the shape of a rendered block, and the GOLDEN.

The golden is test/golden_test.c's own scenario and its own committed constants, driven from
Python. That is the point: it proves the binding reaches the same DSP the C suite pins, so a
binding bug that quietly changed a gain or a position cannot pass. Reproducing it needed no
C-only step - a push source, a position, a pose, a commit and a render loop are all bound.
"""

from __future__ import annotations

import numpy as np
import pytest

import bw_audio as bwa
from conftest import BLK, SR, golden_tone_block

# test/golden_test.c's committed references and its tolerance. Regenerate BOTH together, from that
# test's printout, after an intended DSP change.
G_TOTAL = 168.172397
G_NEGX = 167.502218
G_TOL = 2e-3
NBLK = 48


def _render_golden_scenario():
    """test/golden_test.c's render_scenario, call for call."""
    e = bwa.Engine(profile=bwa.Profile.CAVE, sample_rate=SR, block_size=BLK, sink=bwa.SinkType.MANUAL)
    try:
        nch = e.channel_count
        spk = np.array(e.speakers, copy=True)

        s = e.create_push_source()
        s.set_pos(-1.5, 1.5, 0.0)  # room -x is the listener's RIGHT
        e.listener.set_pose(0.0, 1.5, 0.0, 0.0, 0.0, 0.0, 1.0)
        e.commit()
        e.start()

        energy = np.zeros(nch, dtype=np.float64)
        head = []
        for b in range(NBLK):
            s.push(golden_tone_block(b))
            out = e.render_block()
            assert out is not None and out.shape == (nch, BLK)
            energy += np.sum(out.astype(np.float64) ** 2, axis=1)
            if len(head) < BLK:
                head.append(np.array(out[0], copy=True))

        # The manual sink cannot observe a dropout, and must say so rather than report a clean
        # bill. This is the offline blind spot the counters exist to be honest about.
        assert e.health is None

        total = float(energy.sum())
        neg_x = float(energy[spk[:, 0] < -0.5].sum())
        pos_x = float(energy[spk[:, 0] > 0.5].sum())
        return total, neg_x, pos_x, np.concatenate(head)
    finally:
        e.close()


def test_golden_matches_the_c_suite():
    total, neg_x, pos_x, _ = _render_golden_scenario()
    assert abs(total - G_TOTAL) <= G_TOL * G_TOTAL, (
        "golden total {0:.6f} against {1:.6f}. If the DSP changed on purpose, regenerate this AND "
        "test/golden_test.c's constants from that test's printout.".format(total, G_TOTAL)
    )
    assert abs(neg_x - G_NEGX) <= G_TOL * G_NEGX
    assert neg_x > pos_x * 1.1, "a room -x source must be right-biased"


def test_render_is_reproducible():
    """Two independent renders are bit-identical. That is what makes a committed golden mean anything."""
    a = _render_golden_scenario()
    b = _render_golden_scenario()
    assert a[0] == b[0] and a[1] == b[1] and a[2] == b[2]
    assert np.array_equal(a[3], b[3])


def test_block_shape_dtype_and_read_only(manual_engine):
    e = manual_engine
    out = e.render_block()
    assert out.shape == (e.channel_count, e.block_size)
    assert out.dtype == np.float32
    assert not out.flags.writeable
    with pytest.raises(ValueError):
        out[0, 0] = 1.0


def test_block_is_a_view_that_the_next_render_overwrites(manual_engine):
    """The array aliases the engine's own buffer. Copy it if you keep it - the docstring says so,
    and this is what makes that claim testable rather than a promise."""
    e = manual_engine
    s = e.create_push_source()
    s.set_pos(0.0, 1.5, 1.0)
    e.commit()

    s.push(np.full(e.block_size, 0.5, dtype=np.float32))
    first = e.render_block()
    kept = first.copy()
    assert np.abs(kept).max() > 0.0

    s.push(np.zeros(e.block_size, dtype=np.float32))
    e.render_block()
    assert not np.array_equal(first, kept), "render_block should hand back a view, not a copy"

    # And copy=True survives the next render, which is the whole reason it exists.
    s.push(np.full(e.block_size, 0.5, dtype=np.float32))
    owned = e.render_block(copy=True)
    snapshot = np.array(owned, copy=True)
    s.push(np.zeros(e.block_size, dtype=np.float32))
    e.render_block()
    assert np.array_equal(owned, snapshot)


def test_render_block_returns_none_off_the_manual_sink(null_engine):
    null_engine.start()
    assert null_engine.render_block() is None


def test_binaural_profile_renders_stereo():
    with bwa.Engine(profile=bwa.Profile.BINAURAL, sample_rate=SR, block_size=BLK,
                    sink=bwa.SinkType.MANUAL) as e:
        out = e.render_block()
        assert out.shape == (2, BLK)


def test_bus_levels_track_the_render(manual_engine):
    e = manual_engine
    s = e.create_push_source()
    s.set_pos(-1.5, 1.5, 0.0)
    e.commit()
    for b in range(4):
        s.push(golden_tone_block(b))
        e.render_block()
    peaks = e.bus_levels
    assert peaks.shape == (e.channel_count,)
    assert float(peaks.max()) > 0.0
    assert e.active_voices >= 1

File: bindings/matlab/tests/test_render.m
function test_render()
%TEST_RENDER  The manual sink's block, and the GOLDEN.
%
%   The golden is test/golden_test.c's own scenario and its own committed constants, driven from
%   here. That is the point: it proves this binding reaches the same DSP the C suite pins, so a
%   binding bug that quietly changed a gain, a position or a channel order cannot pass.
%   Reproducing it needed no C-only step: a push source, a position, a pose, a commit and a
%   render loop are all bound.

SR = 48000;
BLK = 256;

% test/golden_test.c's committed references and its tolerance. Regenerate BOTH together, from
% that test's printout, after an intended DSP change.
G_TOTAL = 168.172397;
G_NEGX  = 167.502218;
G_TOL   = 2e-3;

[total, negx, posx, head] = renderGolden(SR, BLK);

tcheck('the golden total matches the C suite', abs(total - G_TOTAL) <= G_TOL * G_TOTAL, ...
       sprintf('%.6f against %.6f', total, G_TOTAL));
tcheck('the golden -x energy matches the C suite', abs(negx - G_NEGX) <= G_TOL * G_NEGX, ...
       sprintf('%.6f against %.6f', negx, G_NEGX));
tcheck('a room -x source is right-biased', negx > posx * 1.1, ...
       sprintf('%.4f against %.4f', negx, posx));

% A committed golden means nothing unless the render is deterministic, so prove it is.
[t2, n2, p2, head2] = renderGolden(SR, BLK);
tcheck('the offline render is bit-identical run to run', ...
       total == t2 && negx == n2 && posx == p2 && isequal(head, head2));

% Shape and class of a block.
C = bwa.Const;
e = bwa.Engine('profile', C.PROFILE_CAVE, 'sink', C.SINK_MANUAL, ...
               'sample_rate', SR, 'block_size', BLK);
e.start();
b = e.render_block();
tcheck('a block is [nframes, channels]', isequal(size(b), [BLK e.channel_count]), ...
       sprintf('%dx%d', size(b,1), size(b,2)));
tcheck('a block is single', isa(b, 'single'));

% It is a COPY, unlike the Python binding's view. That is the documented difference, so pin it:
% the array a caller kept must survive the next render.
s = e.createPushSource();
s.setPos(0, 1.5, 1);
s.push(single(0.5 * ones(BLK, 1)));
first = e.render_block();
kept = first;
tcheck('the first block is not silent', max(abs(first(:))) > 0);
s.push(single(zeros(BLK, 1)));
e.render_block();
tcheck('a returned block is a copy that the next render cannot touch', isequal(first, kept));

% The meters follow the render.
s.push(single(0.5 * ones(BLK, 1)));
e.render_block();
peaks = e.bus_levels;
tcheck('bus_levels is one peak per channel', isequal(size(peaks), [e.channel_count 1]));
tcheck('bus_levels tracks the render', max(peaks) > 0);
tcheck('active_voices sees the push voice', e.active_voices >= 1);
e.close();

% render_block answers [] off the manual sink rather than pretending.
e = bwa.Engine('sink', C.SINK_NULL);
e.start();
tcheck('render_block is [] on a non-manual sink', isempty(e.render_block()));
e.close();
end

function [total, negx, posx, head] = renderGolden(SR, BLK)
% test/golden_test.c's render_scenario, call for call.
NBLK = 48;
C = bwa.Const;
e = bwa.Engine('profile', C.PROFILE_CAVE, 'sink', C.SINK_MANUAL, ...
               'sample_rate', SR, 'block_size', BLK);
cleanup = onCleanup(@() e.close());

nch = e.channel_count;
spk = e.speakers;

s = e.createPushSource();
s.setPos(-1.5, 1.5, 0.0);                       % room -x is the listener's RIGHT
e.listener.setPose(0.0, 1.5, 0.0, 0, 0, 0, 1);
e.commit();
e.start();

energy = zeros(1, nch);
head = [];
for b = 0:(NBLK-1)
    s.push(goldenTone(b, BLK, SR));
    out = e.render_block();
    if ~isequal(size(out), [BLK nch])
        error('bwa:golden', 'render_block returned %dx%d, wanted %dx%d', ...
              size(out,1), size(out,2), BLK, nch);
    end
    energy = energy + sum(double(out) .^ 2, 1);
    if isempty(head)
        head = out(:, 1);
    end
end

% The manual sink cannot observe a dropout, and must say so rather than report a clean bill.
% This is the offline blind spot the counters exist to be honest about.
if ~isempty(e.health)
    error('bwa:golden', 'the manual sink cannot observe a dropout, but health claims it can');
end

total = sum(energy);
negx = sum(energy(spk(:,1) < -0.5));
posx = sum(energy(spk(:,1) > 0.5));
end

function y = goldenTone(block_index, n, SR)
% Exactly test/golden_test.c's fill_tone, in single precision to match its sinf arithmetic.
start = block_index * n;
i = single(start:(start + n - 1));
w = single(6.2831853) * single(440.0);
y = single(0.25) * sin(single(w .* i / single(SR)));
y = single(y(:));
end
